### PR TITLE
chore(codecs): Implement `bytes` framing

### PR DESCRIPTION
**Pull request by pablosichert**
Part of #7463, has been split off from #8236.

### CODE REVIEW

File: src/codecs/framers/bytes.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        Self::new()
+    }
+}
+
+impl Decoder for BytesCodec {
+    type Item = Bytes;
+    type Error = BoxedFramingError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        let frame = src.split_to(src.len());
```
Comment (lukesteensen) [inline src/codecs/framers/bytes.rs:41]: I believe this can just be:
```suggestion
        let frame = src.split();
```

**Review by lukesteensen — APPROVED** [review]

**Comment by pablosichert** [comment]
I discovered this implementation is faulty, it indefinitely loops. Not sure yet how to make it work while also allowing empty frames.

File: src/codecs/framers/bytes.rs (hunk 1 above)
Comment (pablosichert) [reply src/codecs/framers/bytes.rs]: Changed in 7c249bd744d85e8c17fe71d5ad738f0c9508f0d1

**Comment by lukesteensen** [comment]
> I discovered this implementation is faulty, it indefinitely loops. Not sure yet how to make it work while also allowing empty frames.

Ah, does it keep driving `decode` when it returns `Some`? I could see that leading to infinite empty frames. In that case I think it's reasonable to only return frames when we have some new data.

**Comment by pablosichert** [comment]
Yeah, I investigated a bit and found these comments:

https://github.com/tokio-rs/tokio/blob/7e51b44a209dab1de0810d90e3dddff80c2e44a6/tokio-util/src/codec/decoder.rs#L58-L61

https://github.com/tokio-rs/tokio/blob/08ed41f339e345286f26bdbc6c67bf9f2975ed07/tokio-util/src/codec/framed_impl.rs#L123-L151

I tried setting a flag when an empty frame has been emitted and unsetting it once more data arrives, but at the point where `None` is returned on an empty buffer, it will fall back to `decode_eof` internally and at that point we can't distinguish if this is a legitimate new empty frame or from the previous attempt. I could only make it either work with empty frames and always a trailing empty frame, or not at all.

Probably not worth the hassle, since if I see correctly our other codecs also don't support empty frames.

**Pull request merged**